File: day6.hs
import qualified Data.Map as Map
import qualified Data.Set as Set

type Position = (Int, Int)
type Direction = (Int, Int)

getMapRow :: Int -> String -> Map.Map Position Char
getMapRow rowNum row = Map.fromList $ zip (map (, rowNum) [0..]) row

getMap :: String -> Map.Map Position Char
getMap input = Map.unions $ zipWith getMapRow [0..] (lines input)

getStartPos :: Map.Map Position Char -> (Position, Map.Map Position Char)
getStartPos map = (startPos, newMap)
    where
        startPos = head $ Map.keys $ Map.filter (== '^') map
        newMap = Map.insert startPos '.' map

turn :: (Int, Int) -> (Int, Int)
turn tup = (-y, x)
    where
        x = fst tup
        y = snd tup

tupleAdd :: (Int, Int) -> (Int, Int) -> (Int, Int)
tupleAdd tup1 tup2 = (x1 + x2, y1 + y2)
    where
        x1 = fst tup1
        x2 = fst tup2
        y1 = snd tup1
        y2 = snd tup2

walk :: Map.Map Position Char -> Position -> Direction -> Set.Set Position
walk map pos dir = case Map.lookup nextPos map of
    Nothing       -> Set.empty
    Just nextTile -> if nextTile == '#'
                     then walk map pos (turn dir)
                     else Set.insert nextPos $ walk map nextPos dir
    where
        nextPos = tupleAdd pos dir

solve :: String -> Int
solve input = Set.size $ walk map startPos (0, -1)
    where
        startPos = fst $ getStartPos $ getMap input
        map = snd $ getStartPos $ getMap input

hasLoop :: Map.Map Position Char -> Position -> Direction -> Set.Set (Position, Direction) -> Bool
hasLoop map pos dir visited = case Map.lookup nextPos map of
    Nothing       -> False
    Just nextTile -> if nextTile == '#'
                     then Set.member (pos, turn dir) visited ||
                          hasLoop map pos (turn dir) (Set.insert (pos, turn dir) visited)
                     else Set.member (nextPos, dir) visited ||
                          hasLoop map nextPos dir (Set.insert (nextPos, dir) visited)
    where
        nextPos = tupleAdd pos dir
    
solveOne :: Map.Map Position Char -> Position -> Direction -> Position -> Bool
solveOne map pos dir newObstruct = hasLoop (Map.insert newObstruct '#' map) pos dir (Set.singleton (pos, dir))

solve2 :: String -> Int
solve2 input = Set.size $ Set.filter (solveOne map startPos (0, -1)) $ walk map startPos (0, -1)
    where
        startPos = fst $ getStartPos $ getMap input
        map = snd $ getStartPos $ getMap input

main :: IO ()
main = do
    input <- readFile "day6.txt"

    let result = solve input
    print result

    let result2 = solve2 input
    print result2

File: day6.py
import aoc
import sys
from copy import deepcopy


sys.setrecursionlimit(10000)


def format_input(input: str):
    lab_map = []
    start_pos = None

    lines = input.splitlines()
    for y in range(len(lines)):
        new_row = []
        for x in range(len(lines[y])):
            if lines[y][x] == '^':
                start_pos = x, y
                new_row.append('.')
            else:
                new_row.append(lines[y][x])
        lab_map.append(new_row)
    return lab_map, start_pos


def safe_check(matrix, x, y):
    inside = 0 <= y < len(matrix) and 0 <= x < len(matrix[y])
    if inside:
        return matrix[y][x] == '#'
    else:
        return None


def walk(lab_map, x, y, dx, dy, visited):
    next_x = x + dx
    next_y = y + dy
    match safe_check(lab_map, next_x, next_y):
        case True:
            return walk(lab_map, x, y, -dy, dx, visited)
        case False:
            visited.add((next_x, next_y))
            return walk(lab_map, next_x, next_y, dx, dy, visited)
        case None:
            return visited


def solve(lab_map, start_pos):
    return walk(lab_map, *start_pos, 0, -1, {start_pos})


def walk2(lab_map, x, y, dx, dy, visited, debug = False):
    if (((x, y), (dx, dy))) in visited:
        return True
    visited.add(((x, y), (dx, dy)))
    next_x = x + dx
    next_y = y + dy
    match safe_check(lab_map, next_x, next_y):
        case True:
            return walk2(lab_map, x, y, -dy, dx, visited, debug)
        case False:
            return walk2(lab_map, next_x, next_y, dx, dy, visited, debug)
        case None:
            return False


def solve2(lab_map: list[list[str]], start_pos, visited):
    res = 0
    for x, y in visited:
        new_map = deepcopy(lab_map)
        if (x, y) == (3, 6):
            debug = False
        else:
            debug = False
        new_map[y][x] = '#'
        res += walk2(new_map, *start_pos, 0, -1, set(), debug)
    return res


if __name__ == '__main__':
    input = format_input(aoc.get_input('day6.txt'))
    visited = solve(*input)
    print(len(visited))
    res = solve2(*input, visited)
    print(res)